### ISSUE TITLE
Broker does not deliver msgs when client recovers from down-time

### ISSUE THREAD

**Issue by asicoe** [emqtt/emqttd#1216]
#### Environment

- OS: alpine linux 3.5
- Erlang/OTP: erts-8.1
- EMQ: v2.3-beta.2

#### Description

I have two paho v1.1.1 clients.
One that publishes to a particular topic and one that subscribes to that topic.
Both are using  QOS 2.
The subscriber client is using manual acks which means I only send the last PUBCOMP QOS 2 message once it finishes the processing. If something fails during processing the application closes the client and it then starts it again, resubscribing to the same topic. Since I am using a persistent session (clean session flag set to false) I am expecting the same message to be replayed (broker replays the last unacked PUBREL and client callbacks with the original message it saved in it's persistent state) but that is not happening based on the broker and client logs. This results in data loss.

The only settings I am changing in the broker config are:
```
mqtt.session.max_inflight = 1
mqtt.session.retry_interval = 20s (5 sec or 1 sec)
mqtt.mqueue.max_length = 0
```
The rest are defaut.

Log sample for interactions of EMQ with subscriber client with id sub for topic1:
```
listener.ssl.external.acceptors=32
node.max_ets_tables=2097152
node.process_limit=2097152
listener.ws.external.acceptors=16
mqtt.session.retry_interval=5s
log.console.level=debug
mqtt.session.max_inflight=1
node.name=16a988614e41@172.17.0.2
listener.tcp.external.max_clients=1000000
mqtt.mqueue.max_length=0
listener.ssl.external.max_clients=500000
node.max_ports=1048576
listener.tcp.external.acceptors=64
log.console=both
listener.ws.external.max_clients=250000
...
```
```
14:54:23.500 [debug] Client(undefined@x.x.x.x:50513): RECV CONNECT(Q0, R0, D0, ClientId=sub, ProtoName=MQTT, ProtoVsn=4, CleanSess=false, KeepAlive=60, Username=undefined, Password=undefined)
14:54:23.510 [debug] Client(sub@x.x.x.x:50513): SEND CONNACK(Q0, R0, D0, AckFlags=0, RetainCode=0)
14:54:23.512 [debug] Session(sub): Subscribe [{<<"$client/sub">>,[{qos,1}]}]
14:54:23.525 [debug] Client(sub@x.x.x.x:50513): RECV SUBSCRIBE(Q1, R0, D0, PacketId=1, TopicTable=[{<<"topic1">>,2}])
14:54:23.525 [debug] Session(sub): Subscribe [{<<"topic1">>,[{qos,2}]}]
14:54:23.526 [debug] Client(sub@x.x.x.x:50513): SEND SUBACK(Q0, R0, D0, PacketId=1, QosTable=[2])
14:54:27.334 [debug] Client(sub@x.x.x.x:50513): SEND PUBLISH(Q2, R0, D0, Topic=topic1, PacketId=1, Payload=<...>)
14:54:27.339 [debug] Client(sub@x.x.x.x:50513): RECV PUBREC(Q0, R0, D0, PacketId=1)
14:54:27.339 [debug] Client(sub@x.x.x.x:50513): SEND PUBREL(Q1, R0, D0, PacketId=1)
14:54:27.931 [debug] Client(sub@x.x.x.x:50513): RECV DISCONNECT(Q0, R0, D0)
14:54:27.931 [debug] Client(sub@x.x.x.x:50513): Shutdown for normal
14:54:27.932 [debug] Session(sub): Client <0.1628.0> EXIT for normal
14:54:38.821 [debug] Client(undefined@x.x.x.x:50553): RECV CONNECT(Q0, R0, D0, ClientId=sub, ProtoName=MQTT, ProtoVsn=4, CleanSess=false, KeepAlive=60, Username=undefined, Password=undefined)
14:54:38.822 [debug] Session(sub): Resumed by <0.1645.0>
14:54:38.822 [debug] Client(sub@x.x.x.x:50553): SEND CONNACK(Q0, R0, D0, AckFlags=1, RetainCode=0)
14:54:38.823 [debug] Session(sub): Subscribe [{<<"$client/sub">>,[{qos,1}]}]
14:54:38.824 [warning] Session(sub): Duplicated subscribe: $client/sub, qos = 1
14:54:38.839 [debug] Client(sub@x.x.x.x:50553): RECV SUBSCRIBE(Q1, R0, D0, PacketId=1, TopicTable=[{<<"topic1">>,2}])
14:54:38.840 [debug] Session(sub): Subscribe [{<<"topic1">>,[{qos,2}]}]
14:54:38.840 [warning] Session(sub): Duplicated subscribe: topic1, qos = 2
14:54:38.840 [debug] Client(sub@x.x.x.x:50553): SEND SUBACK(Q0, R0, D0, PacketId=1, QosTable=[2])
14:55:38.833 [debug] Client(sub@x.x.x.x:50553): RECV PINGREQ(Q0, R0, D0)
14:55:38.833 [debug] Client(sub@x.x.x.x:50553): SEND PINGRESP(Q0, R0, D0)
14:55:48.297 [debug] Client(sub@x.x.x.x:50553): Shutdown for closed
14:55:48.298 [debug] Session(sub): Client <0.1645.0> EXIT for {shutdown,closed}
```
**Note:**
- The test sends just 1 msg and it never gets processed as expected.
- Notice how SEND PUBREL(Q1, R0, D0, PacketId=1) is only sent once after the initial publish but is never sent again after client sub reconnects and subscribes.
- I am using the docker image to spin up the EMQ broker https://github.com/emqtt/emq_docker.git

**Comment by syberkitten** [comment]
I'm having the same kind of issue where

- client A subscribes to 100 topics
using clean session false.

- client A disconnects

- client B publishes 1 retained message to each of the 100 topics

- client A - reconnects, and does not get any of the messages to the already
subscribed topics.

Show behavior be expected to change if publishing using QoS 1 or 0 ?

**Comment by rahulbhanushali** [comment]
Facing a similar issue with QoS set to 1.

**Comment by HJianBo** [comment]
Hi, @syberkitten The result is right. According to [MQTT-3.3.1-6]:
> When a new subscription is established, the last retained message, if any, on each matching topic name MUST be sent to the subscriber

There is not have new subscription relationship created when the client A reconnect to EMQ
(That was created before client B publish messages)

@rahulbhanushali Could you please provide more details information. Because I can't reproduce that. Thanks!

**Comment by HJianBo** [comment]
Hi, @syberkitten The client A should receive a message each of 100 topics when it reconnects to the server. (But in my test, It works well)

Can you reproduce this issue now? if you can, please open a new issue and provide more details information. i.e:
- What's the client A subscribed topics QoS level?
- What's the client B published message QoS?
- EMQ version?

It's very helpful. Thanks a lot!

**Comment by HJianBo** [comment]
@asicoe This issue has fixed at v2.3.3

Hi, @syberkitten Suggest opening a new issue If has any problem about retained message.
